events/events_test.go

```diff
@@ -0,0 +1,170 @@
+	defer s.e.RemoveListener(l2)
+
+	// emit event by value
+	s.e.EmitEvent(testEvent{})
+
+	wg1.Wait()
+
+	l1.AssertCalled(s.T(), "HandleEvent", testEvent{})
+	l1.AssertNumberOfCalls(s.T(), "HandleEvent", 1)
+
+	// remove listener
+	s.e.RemoveListener(l1)
+
+	// emit new event that should not be emitted to the listener
+	s.e.EmitEvent(testEvent{})
+
+	wg2.Wait()
+
+	l1.AssertCalled(s.T(), "HandleEvent", testEvent{})
+	l1.AssertNumberOfCalls(s.T(), "HandleEvent", 1)
+
+	l2.AssertNumberOfCalls(s.T(), "HandleEvent", 2)
+}
+
+func (s *EventsTestSuite) TestRegisterNilListener() {
+	added := s.e.AddListener(nil)
+	s.Assert().False(added, "expected a nil listener to not be added")
+
+	// this is to test that there is no nil pointer dereference when emitting a
+	// value after a nil listener has been added
+	s.e.EmitEvent(testEvent{})
+}
+
+func (s *EventsTestSuite) TestDeregisterUnregisteredListener() {
+	unregistered := &MockEventListener{}
+	removed := s.e.RemoveListener(unregistered)
+	s.Assert().False(removed, "expected to not remove a listener that never has been added")
+}
+
+func (s *EventsTestSuite) TestDeregisterNilListener() {
+	removed := s.e.RemoveListener(nil)
+	s.Assert().False(removed, "expect a nil listener to not be removed since it has not been added")
+
+	s.e.AddListener(nil)
+
+	removed = s.e.RemoveListener(nil)
+	s.Assert().False(removed, "expect a nil listener to not be removed since it can't be added")
+}
+
+func (s *EventsTestSuite) TestDeregisterDuringHandleEvent() {
+	var wg1 sync.WaitGroup
+
+	// expect listener to be invoked only once
+	wg1.Add(1)
+	l1 := &MockEventListener{}
+	l1.On("HandleEvent", testEvent{}).Run(func(args mock.Arguments) {
+		s.e.RemoveListener(l1)
+		wg1.Done()
+	}).Return()
+	s.e.AddListener(l1)
+	// make sure listener does stick around for next tests
+	defer s.e.RemoveListener(l1)
+
+	s.e.EmitEvent(testEvent{})
+
+	wg1.Wait()
+
+	l1.AssertCalled(s.T(), "HandleEvent", testEvent{})
+	l1.AssertNumberOfCalls(s.T(), "HandleEvent", 1)
+
+	s.e.EmitEvent(testEvent{})
+
+	l1.AssertCalled(s.T(), "HandleEvent", testEvent{})
+	l1.AssertNumberOfCalls(s.T(), "HandleEvent", 1)
+}
+
+func (s *EventsTestSuite) TestRegisterTwice() {
+	var wg1 sync.WaitGroup
+
+	// expect listener to be invoked only once
+	wg1.Add(1)
+	l1 := &MockEventListener{}
+	l1.On("HandleEvent", testEvent{}).Run(func(args mock.Arguments) {
+		wg1.Done()
+	}).Return()
+
+	added := s.e.AddListener(l1)
+	s.Assert().True(added, "expected a listener to be added on the first occurance")
+	added = s.e.AddListener(l1)
+	s.Assert().False(added, "expected a listener to not be added on the second occurance")
+
+	s.e.EmitEvent(testEvent{})
+
+	wg1.Wait()
+
+	l1.AssertCalled(s.T(), "HandleEvent", testEvent{})
+	l1.AssertNumberOfCalls(s.T(), "HandleEvent", 1)
+}
+
+func TestSyncEventEmitter(t *testing.T) {
+	suite.Run(t, &EventsTestSuite{
+		e: &SyncEventEmitter{},
+	})
+}
+
+func TestAsyncEventEmitter(t *testing.T) {
```

Comment:
Can you also add the following tests:

* Emit an event to an empty group (no registered event listeners).
* Make `handleEvent` throw an exception, and verify the subsequent events call the handlers.